types/time.go

```diff
@@ -1687,55 +1718,57 @@ func extractTimeValue(format string, index, cnt int) (int64, int64, int64, float
 		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(originalFmt)
 	}
 
-	hours, err := strconv.ParseFloat(fields[HourIndex], 64)
+	hours, err := strconv.ParseInt(fields[HourIndex], 10, 64)
 	if err != nil {
 		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(originalFmt)
 	}
-	minutes, err := strconv.ParseFloat(fields[MinuteIndex], 64)
+	minutes, err := strconv.ParseInt(fields[MinuteIndex], 10, 64)
 	if err != nil {
 		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(originalFmt)
 	}
-	seconds, err := strconv.ParseFloat(fields[SecondIndex], 64)
+	seconds, err := strconv.ParseInt(fields[SecondIndex], 10, 64)
 	if err != nil {
 		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(originalFmt)
 	}
-	microseconds, err := strconv.ParseFloat(alignFrac(fields[MicrosecondIndex], MaxFsp), 64)
+	microseconds, err := strconv.ParseInt(alignFrac(fields[MicrosecondIndex], MaxFsp), 10, 64)
 	if err != nil {
 		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(originalFmt)
 	}
-	durations := hours*float64(gotime.Hour) + minutes*float64(gotime.Minute) +
-		seconds*float64(gotime.Second) + microseconds*float64(gotime.Microsecond)
-
-	return years, months, days, durations, nil
+	seconds = hours*3600 + minutes*60 + seconds
+	days += seconds / (3600 * 24)
+	seconds %= 3600 * 24
+	return years, months, days, seconds*int64(gotime.Second) + microseconds*int64(gotime.Microsecond), nil
 }
 
-// ExtractTimeValue extracts time value from time unit and format.
-func ExtractTimeValue(unit string, format string) (int64, int64, int64, float64, error) {
+// ParseDurationValue parses time value from time unit and format.
+// Returns y years m months d days + n nanoseconds
+// Nanoseconds will no longer than one day.
+func ParseDurationValue(unit string, format string) (y int64, m int64, d int64, n int64, _ error) {
```

Comment:
Is this time changing necessary?

Reply:
I think Parse is better than Extract since we do not only extract the value to variables, but also do something to avoid overflow.